File: app/src/main/java/com/cainanbt/controleja/data/model/Transaction.kt
package com.cainanbt.controleja.data.model

import com.google.gson.annotations.SerializedName
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

data class Transaction(
    @SerializedName("id") val id: String? = null,
    @SerializedName("name") val name: String = "",
    @SerializedName("description") val description: String? = null,
    @SerializedName("type") val type: TransactionType = TransactionType.DESPESA,
    @SerializedName("amount") val amount: Double = 0.0,
    @SerializedName("date") val date: Long = System.currentTimeMillis(),
    @SerializedName("paid") val paid: Boolean = false,
    @SerializedName("accountId") val accountId: String? = null,
    @SerializedName("targetAccountId") val targetAccountId: String? = null,
    @SerializedName("categoryId") val categoryId: String? = null,
    @SerializedName("categoryName") val categoryName: String? = null,
    @SerializedName("accountName") val accountName: String? = null,
    @SerializedName("isRecurring") val isRecurring: Boolean = false,
    @SerializedName("isFixed") val isFixed: Boolean = false,
    @SerializedName("recurrenceFrequency") val recurrenceFrequency: RecurrenceFrequency? = null,
    @SerializedName("recurrenceEndDate") val recurrenceEndDate: Long? = null,
    @SerializedName("recurrenceRuleId") val recurrenceRuleId: String? = null,
    @SerializedName("installments") val installments: Int = 1,
    @SerializedName("creditCardId") val creditCardId: String? = null,
    @SerializedName("targetInvoiceId") val targetInvoiceId: String? = null,
    @SerializedName("vehicleId") val vehicleId: String? = null,
    @SerializedName("vehicleName") val vehicleName: String? = null,
    @SerializedName("currentOdometer") val currentOdometer: Double? = null,
    @SerializedName("liters") val liters: Double? = null,
    @SerializedName("fuelType") val fuelType: FuelType? = null,
    @SerializedName("efficiency") val efficiency: Double? = null
) {
    private val currencyFormatter = NumberFormat.getCurrencyInstance(Locale("pt", "BR"))
    private val dateFormatter = SimpleDateFormat("dd/MM/yyyy", Locale("pt", "BR"))

    val formattedAmount: String get() = currencyFormatter.format(amount)
    val formattedDate: String get() = dateFormatter.format(Date(date))
    val dateObject: Date get() = Date(date)
    val itemOpacity: Float get() = if (paid) 0.4f else 1.0f
    val isPending: Boolean get() = !paid
    val statusText: String get() = if (paid) "PAGO" else "PENDENTE"
    val canDelete: Boolean get() = categoryName != "Fatura de Cartão"

    val statusColor: Long
        get() = if (paid) 0xFF00E676 else 0xFFFF9800

    val amountColor: Long
        get() = when (type) {
            TransactionType.RECEITA, TransactionType.TRANSFERENCIA_ENTRADA -> 0xFF00E676
            else -> 0xFFFF5252
        }

    val extratoIcon: String
        get() = when (type) {
            TransactionType.RECEITA -> "trending_down"
            TransactionType.DESPESA -> "trending_up"
            TransactionType.TRANSFERENCIA,
            TransactionType.TRANSFERENCIA_ENTRADA,
            TransactionType.TRANSFERENCIA_SAIDA -> "swap_horiz"
            TransactionType.PAGAMENTO_FATURA -> "receipt_long"
        }

    val extratoAmount: String
        get() = when (type) {
            TransactionType.RECEITA, TransactionType.TRANSFERENCIA_ENTRADA -> "+ $formattedAmount"
            else -> "- $formattedAmount"
        }

    val subtitleDisplay: String
        get() {
            val cat = categoryName ?: "Sem categoria"
            val acc = accountName ?: "Sem conta"
            return when (type) {
                TransactionType.TRANSFERENCIA_SAIDA -> "Saiu de: $acc"
                TransactionType.TRANSFERENCIA_ENTRADA -> "Entrou em: $acc"
                else -> "$cat • $acc"
            }
        }
}

data class TransactionGroup(
    val dateHeader: String,
    val transactions: List<Transaction>
)

File: lib/models/transaction.dart
import 'enums.dart';

class Transaction {
  final String? id;
  final String name;
  final String? description;
  final TransactionType type;
  final double amount;
  final int date;
  final bool paid;
  final String? accountId;
  final String? targetAccountId;
  final String? categoryId;
  final String? categoryName;
  final String? accountName;
  final bool isRecurring;
  final bool isFixed;
  final RecurrenceFrequency? recurrenceFrequency;
  final int? recurrenceEndDate;
  final String? recurrenceRuleId;
  final int installments;
  final String? creditCardId;
  final String? targetInvoiceId;
  final String? vehicleId;
  final String? vehicleName;
  final double? currentOdometer;
  final double? liters;
  final FuelType? fuelType;
  final double? efficiency;

  const Transaction({
    this.id,
    this.name = '',
    this.description,
    this.type = TransactionType.DESPESA,
    this.amount = 0,
    this.date = 0,
    this.paid = false,
    this.accountId,
    this.targetAccountId,
    this.categoryId,
    this.categoryName,
    this.accountName,
    this.isRecurring = false,
    this.isFixed = false,
    this.recurrenceFrequency,
    this.recurrenceEndDate,
    this.recurrenceRuleId,
    this.installments = 1,
    this.creditCardId,
    this.targetInvoiceId,
    this.vehicleId,
    this.vehicleName,
    this.currentOdometer,
    this.liters,
    this.fuelType,
    this.efficiency,
  });

  Transaction copyWith({
    String? id,
    String? name,
    String? description,
    TransactionType? type,
    double? amount,
    int? date,
    bool? paid,
    String? accountId,
    String? targetAccountId,
    String? categoryId,
    String? categoryName,
    String? accountName,
    bool? isRecurring,
    bool? isFixed,
    RecurrenceFrequency? recurrenceFrequency,
    int? recurrenceEndDate,
    String? recurrenceRuleId,
    int? installments,
    String? creditCardId,
    String? targetInvoiceId,
    String? vehicleId,
    String? vehicleName,
    double? currentOdometer,
    double? liters,
    FuelType? fuelType,
    double? efficiency,
  }) {
    return Transaction(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      type: type ?? this.type,
      amount: amount ?? this.amount,
      date: date ?? this.date,
      paid: paid ?? this.paid,
      accountId: accountId ?? this.accountId,
      targetAccountId: targetAccountId ?? this.targetAccountId,
      categoryId: categoryId ?? this.categoryId,
      categoryName: categoryName ?? this.categoryName,
      accountName: accountName ?? this.accountName,
      isRecurring: isRecurring ?? this.isRecurring,
      isFixed: isFixed ?? this.isFixed,
      recurrenceFrequency:
          recurrenceFrequency ?? this.recurrenceFrequency,
      recurrenceEndDate: recurrenceEndDate ?? this.recurrenceEndDate,
      recurrenceRuleId: recurrenceRuleId ?? this.recurrenceRuleId,
      installments: installments ?? this.installments,
      creditCardId: creditCardId ?? this.creditCardId,
      targetInvoiceId: targetInvoiceId ?? this.targetInvoiceId,
      vehicleId: vehicleId ?? this.vehicleId,
      vehicleName: vehicleName ?? this.vehicleName,
      currentOdometer: currentOdometer ?? this.currentOdometer,
      liters: liters ?? this.liters,
      fuelType: fuelType ?? this.fuelType,
      efficiency: efficiency ?? this.efficiency,
    );
  }

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String?,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      type: json['type'] != null
          ? TransactionType.fromJson(json['type'] as String)
          : TransactionType.DESPESA,
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      date: json['date'] as int? ?? 0,
      paid: json['paid'] as bool? ?? false,
      accountId: json['accountId'] as String?,
      targetAccountId: json['targetAccountId'] as String?,
      categoryId: json['categoryId'] as String?,
      categoryName: json['categoryName'] as String?,
      accountName: json['accountName'] as String?,
      isRecurring: json['isRecurring'] as bool? ?? false,
      isFixed: json['isFixed'] as bool? ?? false,
      recurrenceFrequency: json['recurrenceFrequency'] != null
          ? RecurrenceFrequency.fromJson(
              json['recurrenceFrequency'] as String)
          : null,
      recurrenceEndDate: json['recurrenceEndDate'] as int?,
      recurrenceRuleId: json['recurrenceRuleId'] as String?,
      installments: json['installments'] as int? ?? 1,
      creditCardId: json['creditCardId'] as String?,
      targetInvoiceId: json['targetInvoiceId'] as String?,
      vehicleId: json['vehicleId'] as String?,
      vehicleName: json['vehicleName'] as String?,
      currentOdometer:
          (json['currentOdometer'] as num?)?.toDouble(),
      liters: (json['liters'] as num?)?.toDouble(),
      fuelType: json['fuelType'] != null
          ? FuelType.fromJson(json['fuelType'] as String)
          : null,
      efficiency: (json['efficiency'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'type': type.toJson(),
      'amount': amount,
      'date': date,
      'paid': paid,
      'accountId': accountId,
      'targetAccountId': targetAccountId,
      'categoryId': categoryId,
      'categoryName': categoryName,
      'accountName': accountName,
      'isRecurring': isRecurring,
      'isFixed': isFixed,
      'recurrenceFrequency': recurrenceFrequency?.toJson(),
      'recurrenceEndDate': recurrenceEndDate,
      'recurrenceRuleId': recurrenceRuleId,
      'installments': installments,
      'creditCardId': creditCardId,
      'targetInvoiceId': targetInvoiceId,
      'vehicleId': vehicleId,
      'vehicleName': vehicleName,
      'currentOdometer': currentOdometer,
      'liters': liters,
      'fuelType': fuelType?.toJson(),
      'efficiency': efficiency,
    };
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Transaction &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'Transaction(id: $id, name: $name, type: $type, amount: $amount, '
      'date: $date, paid: $paid)';
}
